pkg/erofs/erofs.go

```diff
@@ -0,0 +1,643 @@
+}
+
+var squashNotFound = errors.Errorf("erofsfuse program not found")
+
+// erofsFuse - mount squashFile to extractDir
+// return a pointer to the erofsfuse cmd.
+// The caller of the this is responsible for the process created.
+func erofsFuse(squashFile, extractDir string) (*exec.Cmd, error) {
+	var cmd *exec.Cmd
+
+	once.Do(findErofsFuseInfo)
+	if erofsFuseInfo.Path == "" {
+		return cmd, squashNotFound
+	}
+
+	notifyOpts := ""
+	notifyPath := ""
+	if erofsFuseInfo.SupportsNotfiy {
+		sockdir, err := os.MkdirTemp("", "sock")
+		if err != nil {
+			return cmd, err
+		}
+		defer os.RemoveAll(sockdir)
+		notifyPath = filepath.Join(sockdir, "notifypipe")
+		if err := syscall.Mkfifo(notifyPath, 0640); err != nil {
+			return cmd, err
+		}
+		notifyOpts = "notify_pipe=" + notifyPath
+	}
+
+	// given extractDir of path/to/some/dir[/], log to path/to/some/.dir-squashfs.log
+	extractDir = strings.TrimSuffix(extractDir, "/")
+
+	var cmdOut io.Writer
+	var err error
+
+	logf := filepath.Join(path.Dir(extractDir), "."+filepath.Base(extractDir)+"-erofsfuse.log")
+	if cmdOut, err = os.OpenFile(logf, os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0644); err != nil {
+		log.Infof("Failed to open %s for write: %v", logf, err)
+		return cmd, err
+	}
+
+	fiPre, err := os.Lstat(extractDir)
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed stat'ing %q", extractDir)
+	}
+	if fiPre.Mode()&os.ModeSymlink != 0 {
+		return cmd, errors.Errorf("Refusing to mount onto a symbolic linkd")
+	}
+
+	// It would be nice to only enable debug (or maybe to only log to file at all)
+	// if 'stacker --debug', but we do not have access to that info here.
+	// to debug erofsfuse, use "allow_other,debug"
+	optionArgs := "allow_other,debug"
+	if notifyOpts != "" {
+		optionArgs += "," + notifyOpts
+	}
+	cmd = exec.Command(erofsFuseInfo.Path, "-f", "-o", optionArgs, squashFile, extractDir)
+	cmd.Stdin = nil
+	cmd.Stdout = cmdOut
+	cmd.Stderr = cmdOut
+	cmdOut.Write([]byte(fmt.Sprintf("# %s\n", strings.Join(cmd.Args, " "))))
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed writing to %s", logf)
+	}
+	log.Debugf("Extracting %s -> %s with %s [%s]", squashFile, extractDir, erofsFuseInfo.Path, logf)
+	err = cmd.Start()
+	if err != nil {
+		return cmd, err
+	}
+
+	// now poll/wait for one of 3 things to happen
+	// a. child process exits - if it did, then some error has occurred.
+	// b. the directory Entry is different than it was before the call
+	//    to erofsfuse.  We have to do this because we do not have another
+	//    way to know when the mount has been populated.
+	//    https://github.com/vasi/erofsfuse/issues/49
+	// c. a timeout (timeLimit) was hit
+	startTime := time.Now()
+	timeLimit := 30 * time.Second
+	alarmCh := make(chan struct{})
+	go func() {
+		_ = cmd.Wait()
+		close(alarmCh)
+	}()
+	if erofsFuseInfo.SupportsNotfiy {
+		notifyCh := make(chan byte)
+		log.Infof("%s supports notify pipe, watching %q", erofsFuseInfo.Path, notifyPath)
+		go func() {
```

Comment:
Could you comment here on what information is provided via the notify channel? what the return codes/values-read from notify mean?